ext/opentelemetry-ext-wsgi/src/opentelemetry/ext/wsgi/__init__.py

```diff
@@ -24,75 +24,108 @@
 
 from opentelemetry import propagators, trace
 from opentelemetry.ext.wsgi.version import __version__  # noqa
+from opentelemetry.util import time_ns
+
+
+def get_header_from_environ(
+    environ: dict, header_name: str
+) -> typing.List[str]:
+    """Retrieve a HTTP header value from the PEP3333-conforming WSGI environ.
+
+    Returns:
+        A list with a single string with the header value if it exists, else an empty list.
+    """
+    environ_key = "HTTP_" + header_name.upper().replace("-", "_")
+    value = environ.get(environ_key)
+    if value is not None:
+        return [value]
+    return []
+
+
+def add_request_attributes(span, environ):
+    """Adds HTTP request attributes from the PEP3333-conforming WSGI environ to span."""
+
+    span.set_attribute("component", "http")
+    span.set_attribute("http.method", environ["REQUEST_METHOD"])
+
+    host = environ.get("HTTP_HOST")
+    if not host:
+        host = environ["SERVER_NAME"]
+        port = environ["SERVER_PORT"]
+        scheme = environ["wsgi.url_scheme"]
+        if (
+            scheme == "http"
+            and port != "80"
+            or scheme == "https"
+            and port != "443"
+        ):
+            host += ":" + port
+
+    # NOTE: Nonstandard (but see
+    # https://github.com/open-telemetry/opentelemetry-specification/pull/263)
+    span.set_attribute("http.host", host)
+
+    url = environ.get("REQUEST_URI") or environ.get("RAW_URI")
+
+    if url:
+        if url[0] == "/":
+            # We assume that no scheme-relative URLs will be in url here.
+            # After all, if a request is made to http://myserver//foo, we may get
+            # //foo which looks like scheme-relative but isn't.
+            url = environ["wsgi.url_scheme"] + "://" + host + url
+        elif not url.startswith(environ["wsgi.url_scheme"] + ":"):
+            # Something fishy is in RAW_URL. Let's fall back to request_uri()
```

Comment:
Why do you think this is common? Personally, I don't really know how often that occurs in practice,  but I imagine a buggy client could send a line like `GET foobar HTTP/1.0` instead of `GET /foobar HTTP/1.0`.